datainterface/fastener.py

```diff
@@ -0,0 +1,564 @@
+            self.length_n = (
+                Screw.fraction_str_to_float(self.length) * Fastener.INCH_TO_MM
+            )
+            self.diameter_n = (
+                Screw.decimal_str_to_float(self.diameter) * Fastener.INCH_TO_MM
+            )
+            self.length_n = Screw.decimal_str_to_float(self.pitch) * Fastener.INCH_TO_MM
+
+    @staticmethod
+    def fraction_str_to_float(inch: str):
+        return float(Fraction(inch))
+
+    @staticmethod
+    def decimal_str_to_float(mm: str):
+        return float(Decimal(mm))
+
+    @staticmethod
+    def validate_head(prospect: str):
+        return prospect in Screw.VALID_HEAD
+
+    @staticmethod
+    def validate_drive(prospect: str):
+        return prospect in Screw.VALID_DRIVE
+
+    @staticmethod
+    def validate_direction(prospect: str):
+        return prospect in Screw.VALID_DIRECTION
+
+    @staticmethod
+    def validate_finish(prospect: str):
+        return prospect in Screw.VALID_FINISH
+
+    @staticmethod
+    def validate_length(prospect_measurement_system, prospect_length):
+        if prospect_measurement_system == "metric":
+            if Fastener.validate_decimal_str(prospect_length):
+                l = float(Decimal(prospect_length))
+                return l >= Screw.VALID_LENGTH_N[0] and l <= Screw.VALID_LENGTH_N[1]
+        elif prospect_measurement_system == "imperial":
+            if Fastener.validate_fraction_str(prospect_length):
+                l = float(Decimal(prospect_length)) * Fastener.INCH_TO_MM
+                return l >= Screw.VALID_LENGTH_N[0] and l <= Screw.VALID_LENGTH_N[1]
+        return False
+
+    @staticmethod
+    def validate_diameter(prospect_measurement_system, prospect_diameter):
+        if prospect_measurement_system == "metric":
+            if Fastener.validate_decimal_str(prospect_diameter):
+                l = float(Decimal(prospect_diameter))
+                return l >= Screw.VALID_DIAMETER_N[0] and l <= Screw.VALID_DIAMETER_N[1]
+        elif prospect_measurement_system == "imperial":
+            if Fastener.validate_decimal_str(prospect_diameter):
+                l = float(Decimal(prospect_diameter)) * Fastener.INCH_TO_MM
+                return l >= Screw.VALID_DIAMETER_N[0] and l <= Screw.VALID_DIAMETER_N[1]
+        return False
+
+    @staticmethod
+    def validate_pitch(prospect_measurement_system, prospect_pitch):
+        if prospect_measurement_system == "metric":
+            if Fastener.validate_decimal_str(prospect_pitch):
+                l = float(Decimal(prospect_pitch))
+                return l >= Screw.VALID_PITCH_N[0] and l <= Screw.VALID_PITCH_N[1]
+        elif prospect_measurement_system == "imperial":
+            if Fastener.validate_decimal_str(prospect_pitch):
+                l = float(Decimal(prospect_pitch)) * Fastener.INCH_TO_MM
+                return l >= Screw.VALID_PITCH_N[0] and l <= Screw.VALID_PITCH_N[1]
+        return False
+
+
+class Nut(Fastener):
```

Comment:
`Nut` and `Washer` will be brought up to date with `Screw` once we get to genning data for these fasteners?

Reply:
Yes, I left them in to show the proof of concept of using an abstract `Fastener` class.